### PR TITLE
WIP: Core Audio Backend

### PR DESCRIPTION
**Pull request by kfreezen**
I added a basic Driver class for a Core Audio backend, and added a RadioGroupBox to the preference dialog for selecting Core Audio.

Why did I want to add this code/What are the benefits?
I wanted MuseScore to switch output devices automatically if I were to plug one in. PortAudio required me to mess with the I/O tab in the Preferences dialog to switch output devices. PortAudio has no API for tracking output device changes.

Using Core Audio directly has another benefit, which is a lower "Energy Impact" score. My not-so-scientific test yielded a difference between 5 and 2.5 when playback was stopped. It's not groundbreaking, but it is a small improvement.

- [x] I signed [CLA](https://musescore.org/en/cla)
- [ ] I made sure the code in the PR follows [the coding rules](https://musescore.org/en/handbook/developers-handbook/finding-your-way-around/musescore-coding-rules)
- [x] I made sure the code compiles on my machine
- [x] I made sure there are no unnecessary changes in the code
- [x] I made sure the title of the PR reflects the core meaning of the issue you are solving
- [x] I made sure the commit message(s) contain a description and answer the question "Why do those changes fix that particular issue?" or "Why are those changes really necessary as improvements?"
- [x] I made sure the commit message title starts with "fix #424242:" if there is a related issue
- [ ] I created the test (mtest, vtest, script test) to verify the changes I made

### CODE REVIEW

File: audio/drivers/coreaudio.cpp [hunk 1]
```diff
@@ -0,0 +1,409 @@
...
+      AudioObjectPropertyAddress outputDeviceAddress = {
+            kAudioHardwarePropertyDefaultOutputDevice,
+            kAudioObjectPropertyScopeGlobal,
+            kAudioObjectPropertyElementMaster
+      };
+
+      AudioObjectAddPropertyListener(kAudioObjectSystemObject, &outputDeviceAddress, &onDefaultOutputDeviceChanged, this);
+
+      AudioDeviceID deviceId = getDefaultOutputDevice();
+      return openAudioUnit(&audioUnit, &deviceID);
```
Comment (mattmcclinch) [inline audio/drivers/coreaudio.cpp:259]: How does this compile on your machine and on Travis CI? The spelling of `deviceID` does not match the way it is spelled in the above line, and you forgot to pass `this` as the third parameter to `openAudioUnit`.

**Comment by mattmcclinch** [comment]
Thank you for the work you have done and are continuing to do on this. I feel that lack of CoreAudio support has been a huge drawback up until now. But when I use the CoreAudio engine (after applying your patch) the playback pitch of all notes drops by almost a whole step. When I use the PortAudio engine, the notes play at their correct pitches.

File: audio/drivers/coreaudio.cpp [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+	// Starting with 10.6 systems, the HAL notification thread is created internally
+	if (major == 10 && minor >= 6) {
+		CFRunLoopRef theRunLoop = NULL;
+		AudioObjectPropertyAddress theAddress = { kAudioHardwarePropertyRunLoop, kAudioObjectPropertyScopeGlobal, kAudioObjectPropertyElementMaster };
+		OSStatus osErr = AudioObjectSetPropertyData (kAudioObjectSystemObject, &theAddress, 0, NULL, sizeof(CFRunLoopRef), &theRunLoop);
+		if (osErr != noErr) {
+                  initialized = false;
+                  return false;
+		}
+	}
```
Comment (mattmcclinch) [inline audio/drivers/coreaudio.cpp:248]: The lowest version of macOS that MuseScore supports is 10.10. For this reason, and because `Gestalt` is deprecated as of macOS 10.8, we can probably skip the version check and make this whole block unconditional.

File: audio/drivers/coreaudio.cpp [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+AudioUnit audioUnit = NULL;
+
+//---------------------------------------------------------
+//   CoreAudio
+//---------------------------------------------------------
+
+CoreAudio::CoreAudio(Seq* s)
+   : Driver(s)
+      {
+      _sampleRate = 48000;    // will be replaced by device default sample rate
```
Comment (mattmcclinch) [inline audio/drivers/coreaudio.cpp:47]: This needs to be 44100 (44.1 kHz) in order for the notes to sound at the correct pitch, at least with the default soundfont.

File: CMakeLists.txt [hunk 4]
```diff
@@ -188,7 +189,7 @@ if (APPLE)
 
       set(HAS_AUDIOFILE TRUE)            # Requires libsndfile
       set(MAC_APPCAST_URL "")
-
+      set(USE_COREAUDIO 1)
```
Comment (Jojo-Schmitz) [inline CMakeLists.txt:192]: sure about this? Not depending on `BUILD_COREAUDIO`

File: audio/drivers/coreaudio.h [hunk 5]
```diff
@@ -0,0 +1,80 @@
+//=============================================================================
+//  MuseScore
+//  Linux Music Score Editor
+//
+//  Copyright (C) 2002-2009 Werner Schweer and others
```
Comment (Jojo-Schmitz) [inline audio/drivers/coreaudio.h:5]: `2020 MuseScore BVBA` please

File: mscore/prefsdialog.ui [hunk 6]
```diff
@@ -90,7 +90,7 @@
       <string>Preferences Tab Manager</string>
      </property>
      <property name="currentIndex">
-      <number>0</number>
+      <number>4</number>
```
Comment (Jojo-Schmitz) [inline mscore/prefsdialog.ui:93]: unneeded change, even if most probably harmless

File: mscore/prefsdialog.ui [hunk 7]
```diff
@@ -6,7 +6,7 @@
    <rect>
     <x>0</x>
     <y>0</y>
-    <width>834</width>
+    <width>882</width>
```
Comment (Jojo-Schmitz) [inline mscore/prefsdialog.ui:9]: really needed?

File: audio/drivers/coreaudio.cpp [hunk 8]
```diff
@@ -0,0 +1,409 @@
+//=============================================================================
+//  MusE Score
+//  Linux Music Score Editor
+//
+//  Copyright (C) 2002-2010 Werner Schweer and others
```
Comment (Jojo-Schmitz) [inline audio/drivers/coreaudio.cpp:5]: `2020 MuseScore BVBA` please

File: mscore/prefsdialog.ui [hunk 9]
```diff
@@ -2703,6 +2703,25 @@
        <string>I/O</string>
       </attribute>
       <layout class="QVBoxLayout" name="verticalLayout_8">
+       <item>
+        <layout class="QGridLayout" name="gridLayout_16"/>
+       </item>
```
Comment (mattmcclinch) [inline mscore/prefsdialog.ui:2708]: You can remove this.

File: mscore/prefsdialog.ui [hunk 10]
```diff
@@ -2703,6 +2703,25 @@
...
+          </size>
+         </property>
+         <property name="title">
+          <string>Core Audio</string>
+         </property>
+         <property name="checkable">
+          <bool>true</bool>
+         </property>
+        </widget>
+       </item>
```
Comment (mattmcclinch) [inline mscore/prefsdialog.ui:2724]: Qt Creator made all sorts of unnecessary changes to this file. This is all that is needed. That and adding a tabstop for the coreAudioDriver widget towards the end of the file.

**Comment by mattmcclinch** [comment]
@kfreezen, are you still working on this? And are you willing to do what it takes to make this pull request mergeable? If so, then I can give you as much or as little help as you need. Otherwise, I would be happy to take over from here.

**Comment by kfreezen** [comment]
@mattmcclinch Yes, I'm still working on it. I got to the point where I didn't know what else to do, and that I'd add a PR here, and get other people's input on it. As for specific comments on the code, I will be replying to those throughout the day today.

File: audio/drivers/coreaudio.cpp (hunk 1 above)
Comment (kfreezen) [reply audio/drivers/coreaudio.cpp:259]: It doesn't. I made the mistake of refactoring some unneeded code out of the file, and not compiling after I refactored. Apologies for that oversight.

Comment (kfreezen) [reply audio/drivers/coreaudio.cpp:259]: I will address this in a commit today.

**Comment by mattmcclinch** [comment]
By the way, if you don't know what else to do, you can work on reformatting your code so that it follows the [coding rules](https://musescore.org/en/handbook/developers-handbook/finding-your-way-around/musescore-coding-rules).

**Comment by kfreezen** [comment]
@mattmcclinch is there an auto-formatter somewhere I can use for that?

**Comment by mattmcclinch** [comment]
You will probably have to do it by hand.

**Comment by Jojo-Schmitz** [comment]
There is an xml file suitable for QtCreator in the developerts' handbook

File: audio/drivers/driver.cpp [hunk 11]
```diff
@@ -106,6 +109,20 @@ Driver* driverFactory(Seq* seq, QString driverName)
...
+      if (useCoreAudioFlag) {
+        driver = new CoreAudio(seq);
+        if (!driver->init()) {
+          qDebug("init CoreAudio failed");
+          delete driver;
+          driver = 0;
+        } else {
+          coreAudioIsUsed = true;
+        }
+      }
```
Comment (Jojo-Schmitz) [inline audio/drivers/driver.cpp:122]: 6 spaces indent and banner style braces

File: audio/drivers/coreaudio.cpp (hunk 2 above)
Comment (kfreezen) [reply audio/drivers/coreaudio.cpp]: Resolved in badd47

File: CMakeLists.txt (hunk 4 above)
Comment (kfreezen) [reply CMakeLists.txt]: Removed in 8504c0f. Proper use is further down in the file.

File: audio/drivers/coreaudio.cpp (hunk 1 above)
Comment (kfreezen) [reply audio/drivers/coreaudio.cpp]: Fixed in 8d8ed08eb

File: mscore/prefsdialog.ui (hunk 10 above)
Comment (Jojo-Schmitz) [reply mscore/prefsdialog.ui:2721]: that tabstop got lost now

Comment (mattmcclinch) [reply mscore/prefsdialog.ui:2721]: It didn't get lost, it was never added in the first place.

File: audio/drivers/driver.cpp [hunk 12]
```diff
@@ -106,6 +109,20 @@ Driver* driverFactory(Seq* seq, QString driverName)
...
+#else
+      (void)useCoreAudioFlag; // avoid compiler warning
```
Comment (Jojo-Schmitz) [inline audio/drivers/driver.cpp:124]: or `better `Q_UNUSED(useCoreAudioFlag);

File: audio/drivers/driver.cpp [hunk 13]
```diff
@@ -56,6 +58,7 @@ Driver* driverFactory(Seq* seq, QString driverName)
       bool useAlsaFlag       = preferences.getBool(PREF_IO_ALSA_USEALSAAUDIO);
       bool usePortaudioFlag  = preferences.getBool(PREF_IO_PORTAUDIO_USEPORTAUDIO);
       bool usePulseAudioFlag = preferences.getBool(PREF_IO_PULSEAUDIO_USEPULSEAUDIO);
+      bool useCoreAudioFlag = preferences.getBool(PREF_IO_COREAUDIO_USECOREAUDIO);
```
Comment (Jojo-Schmitz) [inline audio/drivers/driver.cpp:61]: adding a space before the = to align code better

File: audio/drivers/coreaudio.cpp [hunk 14]
```diff
@@ -0,0 +1,384 @@
...
+      AudioObjectPropertyAddress outputDeviceAddress = {
+            kAudioHardwarePropertyDefaultOutputDevice,
+            kAudioObjectPropertyScopeGlobal,
+            kAudioObjectPropertyElementMaster
+      };
+
+      AudioObjectAddPropertyListener(kAudioObjectSystemObject, &outputDeviceAddress, &onDefaultOutputDeviceChanged, this);
+
+      AudioDeviceID deviceId = getDefaultOutputDevice();
+      return openAudioUnit(&audioUnit, &deviceId, this);
```
Comment (mattmcclinch) [inline audio/drivers/coreaudio.cpp:237]: You still need to add
```
      _sampleRate = getSampleRate(deviceId);
```
before this return statement in order for notes to sound at the correct pitch.

File: audio/drivers/coreaudio.cpp [hunk 15]
```diff
@@ -0,0 +1,384 @@
...
+                               AudioBufferList *ioData)
+{
+      seq->setInitialMillisecondTimestampWithLatency();
+
+      long bytesPerFrame = sizeof(float) * ioData->mBuffers[0].mNumberChannels;
+
+      float* buffer = static_cast<float*>(ioData->mBuffers[0].mData);
+      UInt64 frames = ioData->mBuffers[0].mDataByteSize / bytesPerFrame;
+
+      seq->process((unsigned)frames, buffer);
```
Comment (mattmcclinch) [inline audio/drivers/coreaudio.cpp:212]: You might want to do
```
      if (seq->driver())
            seq->process((unsigned)frames, buffer);
```
in order to prevent the crash at program exit.

File: audio/drivers/coreaudio.cpp (hunk 14 above)
Comment (kfreezen) [reply audio/drivers/coreaudio.cpp:237]: Ah yes. There's a couple places I need to change for this to work. See lines 167-171. Should `kAudioUnitScope_Input` be `kAudioUnitScope_Output`?

Comment (kfreezen) [reply audio/drivers/coreaudio.cpp:237]: I think desired output format should probably change with every output device change, yes?

Comment (kfreezen) [reply audio/drivers/coreaudio.cpp:253]: 9c91143 adds the `_sampleRate = getSampleRate(deviceId);`

I was not able to duplicate your issue in any consistent and useful way on my machine. What are you running this build on?

Comment (mattmcclinch) [reply audio/drivers/coreaudio.cpp:253]: MacBook Pro (15 inch, 2018), macOS Catalina Version 10.15.4.

**Comment by kfreezen** [comment]
Are you using the MacBook Pro speakers or something else?

**Comment by mattmcclinch** [comment]
The pitch issue occurred when I was just using the MacBook Pro speakers. If I had earphones plugged in or a bluetooth headset connected, it sounded fine.

File: audio/drivers/coreaudio.cpp (hunk 14 above)
Comment (mattmcclinch) [reply audio/drivers/coreaudio.cpp:253]: With your recent changes you can remove the line in question. It is no longer necessary because of what you have done elsewhere.

Comment (kfreezen) [reply audio/drivers/coreaudio.cpp:253]: Ok. Is your pitch issue fixed, or are you still having issues?

Comment (mattmcclinch) [reply audio/drivers/coreaudio.cpp:253]: I am no longer having issues. The issue is fixed. And it is still fixed even after I remove the line that I asked you to add.

File: CMakeLists.txt [hunk 16]
```diff
@@ -188,7 +189,6 @@ if (APPLE)
 
       set(HAS_AUDIOFILE TRUE)            # Requires libsndfile
       set(MAC_APPCAST_URL "")
-
```
Comment (Jojo-Schmitz) [inline CMakeLists.txt:191]: unnecessary change

File: audio/drivers/coreaudio.cpp [hunk 17]
```diff
@@ -0,0 +1,455 @@
...
+      midiDriver  = 0;
+      }
+
+//---------------------------------------------------------
+//   ~CoreAudio
+//---------------------------------------------------------
+
+CoreAudio::~CoreAudio()
+      {
+      if(initialized) {
```
Comment (Jojo-Schmitz) [inline audio/drivers/coreaudio.cpp:63]: space between ìf`and `(`, here ans elsewhere in this file

File: audio/drivers/coreaudio.cpp [hunk 18]
```diff
@@ -0,0 +1,455 @@
...
+AudioDeviceID getDefaultOutputDevice()
+      {
+      AudioDeviceID devId = 0;
+      UInt32 size = sizeof(AudioDeviceID);
+
+      AudioObjectPropertyAddress devAddress = {
+            kAudioHardwarePropertyDefaultOutputDevice,
+            kAudioObjectPropertyScopeGlobal,
+            kAudioObjectPropertyElementMaster
+      };
```
Comment (Jojo-Schmitz) [inline audio/drivers/coreaudio.cpp:81]: banner style for the closing brace, here end elsewhere in this file

File: audio/drivers/coreaudio.cpp [hunk 19]
```diff
@@ -0,0 +1,455 @@
...
+            return 0.0f;
+
+      }
+
+static OSStatus CoreAudioIOProc( void *inRefCon,
+      AudioUnitRenderActionFlags *ioActionFlags,
+      const AudioTimeStamp *inTimeStamp,
+      UInt32 inBusNumber,
+      UInt32 inNumberFrames,
+      AudioBufferList *ioData );
```
Comment (Jojo-Schmitz) [inline audio/drivers/coreaudio.cpp:112]: the `*` belongs to the type, not the parameter

File: audio/drivers/coreaudio.cpp [hunk 20]
```diff
@@ -0,0 +1,455 @@
...
+      desiredFormat.mFramesPerPacket = 1;
+      desiredFormat.mBitsPerChannel = sizeof(float) * 8;
+      desiredFormat.mSampleRate = sampleRate;
+      desiredFormat.mBytesPerPacket = sizeof(float) * 2;
+      desiredFormat.mBytesPerFrame = sizeof(float) * 2;
+      desiredFormat.mChannelsPerFrame = 2;
+      return AudioUnitSetProperty(*audioUnit, kAudioUnitProperty_StreamFormat, kAudioUnitScope_Input, 0, &desiredFormat, sizeof(desiredFormat));
+            
+      }
+// Substantial portions of this taken from Portaudio.
```
Comment (Jojo-Schmitz) [inline audio/drivers/coreaudio.cpp:132]: add empty line here please

File: audio/drivers/coreaudio.cpp [hunk 21]
```diff
@@ -0,0 +1,455 @@
...
+      desiredFormat.mFormatID = kAudioFormatLinearPCM;
+      desiredFormat.mFormatFlags = kAudioFormatFlagsNativeFloatPacked;
+      desiredFormat.mFramesPerPacket = 1;
+      desiredFormat.mBitsPerChannel = sizeof(float) * 8;
+      desiredFormat.mSampleRate = sampleRate;
+      desiredFormat.mBytesPerPacket = sizeof(float) * 2;
+      desiredFormat.mBytesPerFrame = sizeof(float) * 2;
+      desiredFormat.mChannelsPerFrame = 2;
+      return AudioUnitSetProperty(*audioUnit, kAudioUnitProperty_StreamFormat, kAudioUnitScope_Input, 0, &desiredFormat, sizeof(desiredFormat));
+
```
Comment (Jojo-Schmitz) [inline audio/drivers/coreaudio.cpp:130]: and remove this empty line

File: audio/drivers/coreaudio.cpp [hunk 22]
```diff
@@ -0,0 +1,455 @@
...
+// FIXME: Do we want to keep this around? Will there be a need to give people the option to switch devices in the CoreAudio back end?
+QStringList CoreAudio::deviceList(int apiIdx) {
+  QStringList dl;
+
+  AudioObjectPropertyAddress deviceListAddress = {
+    kAudioHardwarePropertyDevices,
+    kAudioObjectPropertyScopeGlobal,
+    kAudioObjectPropertyElementMaster
+  };
```
Comment (Jojo-Schmitz) [inline audio/drivers/coreaudio.cpp:318]: indent 6 spaces

File: audio/drivers/driver.cpp [hunk 23]
```diff
@@ -90,7 +95,7 @@ Driver* driverFactory(Seq* seq, QString driverName)
                   pulseAudioIsUsed = true;
             }
 #else
-      (void)usePulseAudioFlag; // avoid compiler warning
+      Q_UNUSED(usePulseAudioFlag); // avoid compiler warning
```
Comment (Jojo-Schmitz) [inline audio/drivers/driver.cpp:98]: The comment doesn't harm, but isn't needed anymore

File: mscore/prefsdialog.ui (hunk 10 above)
Comment (Jojo-Schmitz) [reply mscore/prefsdialog.ui:2721]: Can you add it then?

**Comment by Jojo-Schmitz** [comment]
Please squash all these commits into just one

**Comment by kfreezen** [comment]
I'm hoping that covers all the formatting changes. Tried to make sure I got them all.

There seems to be a glitch I'm running into intermittently while using the CoreAudio driver. The position indicator froze at position 0 in the score and the sequencer buffer filled to max. I can't get that to reproduce consistently for me though.

File: audio/drivers/coreaudio.cpp [hunk 24]
```diff
@@ -0,0 +1,466 @@
...
+      char* buffer = static_cast<char*>(malloc(maxSize));
+      if (CFStringGetCString(str, buffer, maxSize, kCFStringEncodingUTF8))
+            return buffer;
+      
+      free(buffer);
+      return nullptr;
+      }
+
+// FIXME: Do we want to keep this around? Will there be a need to give people the option to switch devices in the CoreAudio back end?
+QStringList CoreAudio::deviceList(int apiIdx)
```
Comment (mattmcclinch) [inline audio/drivers/coreaudio.cpp:321]: `apiIdx` does not apply, since CoreAudio only uses the one API.

File: audio/drivers/coreaudio.h [hunk 25]
```diff
@@ -0,0 +1,85 @@
...
+      float* getRBuffer(long n);
+      virtual bool isRealtime() const   { return false; }
+
+      QStringList apiList() const;
+      QStringList deviceList(int apiIdx);
+      int deviceIndex(int apiIdx, int apiDevIdx);
+      int currentApi() const;
+      int currentDevice() const;
+      MidiDriver* mididriver() { return midiDriver; }
+};
```
Comment (mattmcclinch) [inline audio/drivers/coreaudio.h:81]: This closing brace should be indented.

File: audio/drivers/coreaudio.cpp [hunk 26]
```diff
@@ -0,0 +1,466 @@
...
+
+      if (seq->driver())
+            seq->process((unsigned)frames, buffer);
+
+      return noErr;
+      }
+
+bool CoreAudio::init(bool)
+      {
+
```
Comment (mattmcclinch) [inline audio/drivers/coreaudio.cpp:233]: This blank line is not needed.

File: audio/drivers/coreaudio.cpp [hunk 27]
```diff
@@ -0,0 +1,466 @@
...
+      }
+
+bool CoreAudio::start(bool)
+      {
+      OSStatus status = AudioOutputUnitStart(audioUnit);
+      if (status)
+            return false;
+      else
+            return true;
+
```
Comment (mattmcclinch) [inline audio/drivers/coreaudio.cpp:271]: This blank line is not needed.

File: audio/drivers/coreaudio.cpp [hunk 28]
```diff
@@ -0,0 +1,466 @@
...
+bool CoreAudio::stop()
+      {
+      OSStatus status = AudioOutputUnitStop(audioUnit);
+      
+      if (status)
+            return false;
+      else
+            return true;
+
```
Comment (mattmcclinch) [inline audio/drivers/coreaudio.cpp:282]: This blank line is not needed.

File: audio/drivers/coreaudio.h [hunk 29]
```diff
@@ -0,0 +1,85 @@
...
+      
+      int framePos() const;
+      float* getLBuffer(long n);
+      float* getRBuffer(long n);
+      virtual bool isRealtime() const   { return false; }
+
+      QStringList apiList() const;
+      QStringList deviceList(int apiIdx);
+      int deviceIndex(int apiIdx, int apiDevIdx);
+      int currentApi() const;
```
Comment (mattmcclinch) [inline audio/drivers/coreaudio.h:78]: Again, CoreAudio only uses the one API.

File: CMakeLists.txt [hunk 30]
```diff
@@ -499,14 +500,24 @@ ELSE(BUILD_JACK)
...
+##
+## CoreAudio
+##
+
+if (BUILD_COREAUDIO)
+    if (APPLE)
+        message("Including CoreAudio support.")
+        set(USE_COREAUDIO 1)
+    endif (APPLE)
+endif (APPLE)
```
Comment (mattmcclinch) [inline CMakeLists.txt:512]: This should be
```
endif (BUILD_COREAUDIO)
```

File: audio/drivers/coreaudio.cpp [hunk 31]
```diff
@@ -0,0 +1,466 @@
...
+            kAudioObjectPropertyElementMaster
+            };
+
+      OSStatus status = AudioObjectGetPropertyData(devId, &theAddress, 0, NULL, &size, &sampleRate);
+
+      if (status == noErr)
+            return sampleRate;
+      else
+            return 0.0f;
+
```
Comment (Jojo-Schmitz) [inline audio/drivers/coreaudio.cpp:105]: superfluos empty line

File: audio/drivers/coreaudio.cpp [hunk 32]
```diff
@@ -0,0 +1,466 @@
...
+static OSStatus CoreAudioIOProc( void* inRefCon,
+                               AudioUnitRenderActionFlags* ioActionFlags,
+                               const AudioTimeStamp* inTimeStamp,
+                               UInt32 inBusNumber,
+                               UInt32 inNumberFrames,
+                               AudioBufferList* ioData)
+      {
+      seq->setInitialMillisecondTimestampWithLatency();
+
+      long bytesPerFrame = sizeof(float) * ioData->mBuffers[0].mNumberChannels;
```
Comment (Jojo-Schmitz) [inline audio/drivers/coreaudio.cpp:220]: as sizeof is an operator and keyword, please add a space `sizeof (float)`

File: audio/drivers/coreaudio.cpp [hunk 33]
```diff
@@ -0,0 +1,466 @@
...
+
+      float* buffer = static_cast<float*>(ioData->mBuffers[0].mData);
+      UInt64 frames = ioData->mBuffers[0].mDataByteSize / bytesPerFrame;
+
+      if (seq->driver())
+            seq->process((unsigned)frames, buffer);
```
Comment (Jojo-Schmitz) [inline audio/drivers/coreaudio.cpp:226]: please avoid C-style casts. This here may no be needed at all, `frames` being unsigned already?

File: audio/drivers/coreaudio.cpp [hunk 34]
```diff
@@ -0,0 +1,466 @@
...
+      CFMutableArrayRef inputDeviceArray = CFArrayCreateMutable(kCFAllocatorDefault, deviceCount, &kCFTypeArrayCallBacks);
+      if (NULL == inputDeviceArray) {
+            qDebug("CFArrayCreateMutable failed");
+            free(audioDevices), audioDevices = nullptr;
+            return dl;
+            }
+
+      // Iterate through all the devices and determine which are input-capable
+      deviceListAddress.mScope = kAudioDevicePropertyScopeOutput;
+      for(UInt32 i = 0; i < deviceCount; i++) {
```
Comment (Jojo-Schmitz) [inline audio/drivers/coreaudio.cpp:364]: space between `for` and `(`

File: audio/drivers/coreaudio.cpp [hunk 35]
```diff
@@ -0,0 +1,466 @@
...
+                               UInt32 inBusNumber,
+                               UInt32 inNumberFrames,
+                               AudioBufferList* ioData)
+      {
+      seq->setInitialMillisecondTimestampWithLatency();
+
+      long bytesPerFrame = sizeof(float) * ioData->mBuffers[0].mNumberChannels;
+
+      float* buffer = static_cast<float*>(ioData->mBuffers[0].mData);
+      UInt64 frames = ioData->mBuffers[0].mDataByteSize / bytesPerFrame;
```
Comment (Jojo-Schmitz) [inline audio/drivers/coreaudio.cpp:223]: is `UInt64` really needed here?

Comment (Jojo-Schmitz) [reply audio/drivers/coreaudio.cpp:223]: as per https://developer.apple.com/documentation/coreaudiotypes/audiobuffer/1422352-mdatabytesize it is an `UInt32`

File: audio/drivers/coreaudio.cpp (hunk 32 above)
Comment (Jojo-Schmitz) [reply audio/drivers/coreaudio.cpp:220]: sizeof() BTW returns a size_t, which is an unsigned value

File: CMakeLists.txt [hunk 36]
```diff
@@ -188,7 +189,7 @@ if (APPLE)
 
       set(HAS_AUDIOFILE TRUE)            # Requires libsndfile
       set(MAC_APPCAST_URL "")
-
+
```
[the remark this answers is not in the document]
Comment (Jojo-Schmitz) [reply CMakeLists.txt:192]: Still unadressed...

File: CMakeLists.txt [hunk 37]
```diff
@@ -499,14 +500,24 @@ ELSE(BUILD_JACK)
...
+endif (APPLE)
 
 ##
 ## PortAudio
 ##
 
 if (BUILD_PORTAUDIO)
     if (MINGW OR MSVC)
-        set ( USE_PORTAUDIO 1 )
+      set ( USE_PORTAUDIO 1 )
```
[the remark this answers is not in the document]
Comment (Jojo-Schmitz) [reply CMakeLists.txt:520]: still unadressed

File: audio/drivers/coreaudio.h (hunk 25 above)
Comment (Jojo-Schmitz) [reply audio/drivers/coreaudio.h:79]: still unadressed

File: audio/drivers/driver.cpp [hunk 38]
```diff
@@ -33,14 +35,16 @@
 #ifdef USE_PORTAUDIO
 #include "pa.h"
 #endif
-
+#ifdef USE_COREAUDIO
+#include "coreaudio.h"
+#endif
 namespace Ms {
```
[the remark this answers is not in the document]
Comment (Jojo-Schmitz) [reply audio/drivers/driver.cpp:41]: sill unaddressed

File: audio/drivers/coreaudio.cpp [hunk 39]
```diff
@@ -0,0 +1,466 @@
...
+      if (status)
+            return false;
+      else
+            return true;
+
+      }
+
+void CoreAudio::stopTransport()
+      {
+    state = Transport::STOP;
```
[the remark this answers is not in the document]
Comment (Jojo-Schmitz) [reply audio/drivers/coreaudio.cpp:283]: still anaddressed

**Comment by kfreezen** [comment]
I'm curious whether there's anything I missed on this one, or if it's waiting on someone else?

**Comment by Jojo-Schmitz** [comment]
It needs a rebase, adjusting to the new coding style.

**Comment by Jojo-Schmitz** [comment]
It needs a rebase, adjusting to the new coding style. There's a script to help with that, in tools/codestyle.
Also some of the files may have been moved around, quite some changes have been done reorganizing the code.

**Comment by kfreezen** [comment]
There are some glitches that a friend and I have run into while testing this branch over the last 3-4 months.

The main glitch is what looks like the sequencer freezing after the program has been running in the background for a while. I have not narrowed down what causes the bug, since it's fairly intermittent and shows up when you hit the Play button.

It seems like the problem is fixed if I edit the score before playing, however. Maybe someone can give me some ideas on where to look for this one?

**Comment by kfreezen** [comment]
I'm curious what architecture changes are being actively made in MuseScore. Is there a source or documentation on those arch changes? Is audio driver code moving to another project? That's what it looks like.

**Comment by RobFog** [comment]
@igorkorsukov should be able to answer that.

**Comment by igorkorsukov** [comment]
> I'm curious what architecture changes are being actively made in MuseScore. Is there a source or documentation on those arch changes? Is audio driver code moving to another project? That's what it looks like.

Yes, that's right, the architecture of the project has changed a lot.
In general, you can read about the principles of the new project architecture here (not about audio)
https://github.com/musescore/Documentation/blob/master/Architecture/README.md

The new audio driver is here, you can preview it.
https://github.com/musescore/MuseScore/blob/master/src/framework/audio/internal/platform/osx/osxaudiodriver.cpp
 
I am very sorry that we were not able to take advantage of your work.

**Comment by kfreezen** [comment]
@igorkorsukov no problem. As long as I can get "instant switching" with my Airpods, I'm happy. Where is a good place to ask questions about the general architecture and MuseScore's long term goals with it?

**Comment by igorkorsukov** [comment]
> @igorkorsukov no problem. As long as I can get "instant switching" with my Airpods, I'm happy. Where is a good place to ask questions about the general architecture and MuseScore's long term goals with it?

You can ask questions in pull requests or in the development chat https://t.me/musescoreeditorchat (if the question is about architecture, then it's better to mention me)

**Comment by igorkorsukov** [comment]
> As long as I can get "instant switching" with my Airpods, I'm happy.

@psmokotnin FYI
